src/utils/file_upload.rs

```diff
@@ -376,6 +349,66 @@ fn upload_files_chunked(
     Ok(())
 }
 
+fn upload_files_chunked(
+    context: &UploadContext,
+    files: &SourceFiles,
+    options: &ChunkUploadOptions,
+) -> Result<()> {
+    let archive = build_artifact_bundle(context, files, None)?;
+
+    let progress_style =
+        ProgressStyle::default_spinner().template("{spinner} Optimizing bundle for upload...");
+
+    let pb = ProgressBar::new_spinner();
+    pb.enable_steady_tick(100);
+    pb.set_style(progress_style);
+
+    let view = ByteView::open(archive.path())?;
+    let (checksum, checksums) = get_sha1_checksums(&view, options.chunk_size)?;
+    let mut chunks = view
+        .chunks(options.chunk_size as usize)
+        .zip(checksums.iter())
+        .map(|(data, checksum)| Chunk((*checksum, data)))
+        .collect::<Vec<_>>();
+
+    pb.finish_with_duration("Optimizing");
+
+    let progress_style = ProgressStyle::default_bar().template(&format!(
+        "{} Uploading files...\
+       \n{{wide_bar}}  {{bytes}}/{{total_bytes}} ({{eta}})",
+        style(">").dim(),
+    ));
+
+    // Filter out chunks that are already on the server. This only matters if we're in the
+    // `assemble_artifact_bundle` case; `assemble_release_artifacts` always returns `missing_chunks: []`,
+    // so there's no point querying the endpoint.
+    if options.supports(ChunkUploadCapability::ArtifactBundles) && context.project.is_some() {
```

Comment:
Probably not worth it, but we might want to extract this (and another use above) into `use_artifact_bundle` or similar helper, to make sure it's always in sync in all places. But it's only repeated twice right now, so no need to move it yet imo.